arangod/Pregel/Algos/AIR/Greenspun/Primitives.cpp

```diff
@@ -0,0 +1,1264 @@
+      }
+    }
+  };
+
+  auto reduceArray = [&](auto& inputValue) -> EvalResult {
+    for (VPackArrayIterator iter(inputValue); iter.valid(); iter++) {
+      VPackBuilder parameter;
+      buildLambdaParameters(parameter, iter);
+
+      result.clear();
+      EvalResult res = EvaluateApply(ctx, functionSlice,
+                                     VPackArrayIterator(parameter.slice()), result, false);
+      if (res.fail()) {
+        return res.error().wrapMessage("when reducing array parameters " +
+                                       parameter.toJson());
+      }
+    }
+    // no error
+    return {};
+  };
+
+  auto reduceObject = [&](auto& inputValue) -> EvalResult {
+    for (VPackObjectIterator iter(inputValue); iter.valid(); iter++) {
+      VPackBuilder parameter;
+      buildLambdaParameters(parameter, iter);
+
+      result.clear();
+      auto res = EvaluateApply(ctx, functionSlice,
+                               VPackArrayIterator(parameter.slice()), result, false);
+      if (res.fail()) {
+        return res.error().wrapMessage("when reducing object parameters " +
+                                       parameter.toJson());
+      }
+    }
+    // no error
+    return {};
+  };
+
+  if (inputValue.isArray()) {
+    auto res = reduceArray(inputValue);
+    if (res.fail()) {
+      return res;
+    }
+  } else if (inputValue.isObject()) {
+    auto res = reduceObject(inputValue);
+    if (res.fail()) {
+      return res;
+    }
+  } else {
+    return EvalError("expected either object or array as input value, found: " +
+                     inputValue.toJson() +
+                     ". Accumulator can be any type: " + inputAccumulator.toJson() +
+                     " (depends on lambda definition");
+  }
+
+  return {};
+}
+
+EvalResult Prim_Filter(Machine& ctx, VPackSlice const paramsList, VPackBuilder& result) {
+  if (!paramsList.isArray() || paramsList.length() != 2) {
+    return EvalError(
+        "expecting two arguments, a function and a list or object");
+  }
+
+  auto functionSlice = paramsList.at(0);
+  auto list = paramsList.at(1);
+
+  if (list.isArray()) {
+    VPackArrayBuilder ab(&result);
+    for (VPackArrayIterator iter(list); iter.valid(); iter++) {
+      VPackBuilder parameter;
+      {
+        VPackArrayBuilder pb(&parameter);
+        parameter.add(VPackValue(iter.index()));
+        parameter.add(*iter);
+      }
+
+      VPackBuilder filterResult;
+      auto res = EvaluateApply(ctx, functionSlice,
+                               VPackArrayIterator(parameter.slice()), filterResult, false);
+      if (res.fail()) {
+        return res.error().wrapMessage("when filtering pair " + parameter.toJson());
+      }
+
+      if (ValueConsideredTrue(filterResult.slice())) {
+        result.add(*iter);
+      }
+    }
+  } else if (list.isObject()) {
+    VPackObjectBuilder ob(&result);
+    for (VPackObjectIterator iter(list); iter.valid(); iter++) {
```

Comment:
VPackArrayIterator is not completely trivial. We should use preincrement on it.
```suggestion
    for (VPackObjectIterator iter(list); iter.valid(); ++iter) {
```